hardhat/tasks/misc.js

```diff
@@ -268,3 +215,122 @@ task("getBalanceOfAccounts", "Get ERC20 or native token balance of account or ad
+    const staking = await ethers.getContractAt(
+        "IStaking",
+        "0x5684a06CaB22Db16d901fEe2A5C081b4C91eA40e"
+    );
+    //const abi = (await deployments.getArtifact("Staking")).abi;
+    const abi = ["event VestingStakeSet(uint256,uint96)"];
+    //const abi = ["event TokensStaked(address,uint256,uint256,uint256)"];
+    const iface = new ethers.utils.Interface(abi);
+    //const filter = staking.filters.VestingStakeSet(null, null);
+    //cblock = 3780053; // a block with the first vesting created
+    cblock = 5190053; // a block with the first vesting created
+    block = await ethers.provider.getBlockNumber();
+    let index = 0;
+    while (cblock != block) {
+        cblock += 10000;
+        if (cblock > block) cblock = block;
+        const filter = {
+            address: "0x5684a06CaB22Db16d901fEe2A5C081b4C91eA40e",
+            topics: [
+                //ethers.utils.id("TokensStaked(address,uint256,uint256,uint256)")
+                ethers.utils.id("VestingStakeSet(uint256,uint96)"),
+            ],
+            fromBlock: cblock - 10000,
+            toBlock: cblock,
+        };
+        let cres = [];
+
+        try {
+            cres = await ethers.provider.getLogs(filter);
+        } catch (e) {
+            console.log(e);
+            console.log("failure at block", cblock);
+            return;
+        }
+        //console.log(await getEthersLog(staking, filter));
+        if (cres[0]) {
+            console.log("index: ", index++, "\n", cres);
+            //break;
+        }
+        if (cblock % 500000 == 0) {
+            console.log(cblock, "block reached");
+        }
+    }
+});
+
+/**
+ * This is the infura node version to run an eth node
+ * It is throwing the error:
+ * 
+ * "Error HH604: Error running JSON-RPC server: Too Many Requests error received from mainnet.infura.io"
+ * 
+task("misc:run-forked-eth-mainnet", "Runs a forked eth node")
+    .addOptionalParam("forkBlock", "Block number to fork from")
+    .setAction(async (taskArgs, hre) => {
+        const { NODE_REAL_API_KEY } = process.env;
+        const forkParams = {
+            jsonRpcUrl: `https://eth-mainnet.nodereal.io/v1/${NODE_REAL_API_KEY}`,
+            blockNumber: taskArgs.forkBlock ? parseInt(taskArgs.forkBlock) : undefined,
+        };
+        await hre.network.provider.request({
+            method: "hardhat_reset",
+            params: [
+                {
+                    forking: forkParams,
+                },
+            ],
+        });
+        await hre.run("node", { noDeploy: true });
+    }
+);
+*/
+
+// Please obtain a node-real api key and config it on .env file
+task("misc:run-forked-eth-mainnet", "Runs a forked eth node")
+    .addOptionalParam("forkBlock", "Block number to fork from")
+    .setAction(async (taskArgs, hre) => {
+        const { NODE_REAL_API_KEY } = process.env;
+        const forkParams = {
+            jsonRpcUrl: `https://eth-mainnet.nodereal.io/v1/${NODE_REAL_API_KEY}`,
```

Comment:
probably it's a good idea to put all of the fork rpc base url to a config. So that it can be reused, and easier to update.
Something like

const forkRpcUrls = {
"ethMainnet": "https:xxx",
"bobMainnet": "https:xxx",
....
}

Reply:
Solved